### PR TITLE
refactor: move background detection into Lua

### PR DESCRIPTION
**Pull request by gpanders**
Now that we have the `TermResponse` event we have everything we need to do this in Lua instead of C. The Lua implementation is much more straightforward and simple.

There is a functional change: the background detection happens later in the Nvim startup process. This shouldn't matter, as in either case the `'background'` option isn't changed if it has already been set by the user.

This deprecates the `term_background` UI option. I am not removing it in this PR but I am marking it deprecated to be removed in a future release.

### CODE REVIEW

File: runtime/lua/vim/_editor.lua [hunk 1]
```diff
@@ -1176,9 +1176,85 @@ function vim._init_default_autocmds()
...
+--- ignored in the calculations.
+--- 
+--- [1] https://en.wikipedia.org/wiki/Luma_%28video%29
+local function guess_background_color()
+  if vim.api.nvim_get_option_info2('bg', {}).was_set then
+    -- Don't do anything if 'bg' is already set
+    return
+  end
+
+  vim.api.nvim_create_autocmd('TermResponse', {
```
Comment (gpanders) [inline runtime/lua/vim/_editor.lua:1219]: TODO: Add a timer to delete this autocommand if no response is received after some timeout

File: runtime/lua/vim/_editor.lua [hunk 2]
```diff
@@ -1176,9 +1176,85 @@ function vim._init_default_autocmds()
...
+        if bg ~= vim.o.bg then
+          vim.o.bg = bg
+        end
+
+        return true
+      end
+    end,
+  })
+
+  io.stdout:write('\x1b]11;?\x1b\\')
```
Comment (zeertzjq) [inline runtime/lua/vim/_editor.lua:1245]: This doesn't work for remote TUI.

File: runtime/lua/vim/_editor.lua [hunk 3]
```diff
@@ -1176,9 +1176,85 @@ function vim._init_default_autocmds()
...
+end
+
 function vim._init_defaults()
   vim._init_default_mappings()
   vim._init_default_autocmds()
+
+  for _, ui in ipairs(vim.api.nvim_list_uis()) do
+    if ui.stdout_tty then
```
Comment (zeertzjq) [inline runtime/lua/vim/_editor.lua:1253]:
```suggestion
    if ui.chan == 1 and ui.stdout_tty then
```

File: runtime/lua/vim/_editor.lua [hunk 4]
```diff
@@ -1176,9 +1176,97 @@ function vim._init_default_autocmds()
...
+--- fractional value between 0 and 1 representing the intensity of the color in
+--- the range.
+---
+--- For instance, if only a single hex char "a" is used, then this function
+--- returns 0.625 (10 / 16), while a value of "aa" would return 0.664 (170 /
+--- 256).
+---
+--- @param c string Color as a string of hex chars
+--- @return number Intensity of the color
+local function hexcolor(c)
```
Comment (clason) [inline runtime/lua/vim/_editor.lua:1191]:
```suggestion
local function luminance(c)
```
to match the description below? (should use the same term in all docstrings and variable names)

**Review by justinmk — APPROVED** [review]
Awesome!

File: runtime/doc/news.txt [hunk 5]
```diff
@@ -340,4 +340,8 @@ release.
 • vim.treesitter.languagetree functions:
   - |LanguageTree:for_each_child()|	Use |LanguageTree:children()| (non-recursive) instead.
 
+• The "term_background" UI option |ui-ext-options| is deprecated and no longer
```
Comment (justinmk) [inline runtime/doc/news.txt:343]: also worth mentioning in deprecated.txt

**Comment by xdBronch** [comment]
after updating nvim recently i noticed that when running `:wq` on zig files with [zls](https://github.com/zigtools/zls) theres a large delay, 1-2 seconds, before it actually exits. `:w` and `:q` on their own dont do this and ive bisected it to 22eb2ba18336df6cd70a88f666818ee5d8ba92d2, any ideas?

**Comment by zeertzjq** [comment]
Please provide compile flags from `:verbose version`

**Comment by clason** [comment]
Please open a new issue and provide all the necessary information.

**Pull request merged**